%YAML 1.2
# SPDX-FileCopyrightText: NVIDIA CORPORATION & AFFILIATES
# Copyright (c) 2022-2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0
---
name: sync
components:
- name: left_image_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: left_image_receiver
    min_size: 1
- name: right_image_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: right_image_receiver
    min_size: 1
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 12
    policy: 0
- name: left_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: left_cam_receiver
    min_size: 1
- name: right_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: right_cam_receiver
    min_size: 1
- type: nvidia::isaac_ros::Synchronization
  parameters:
    left_image_input: left_image_receiver
    right_image_input: right_image_receiver
    left_cam_input: left_cam_receiver
    right_cam_input: right_cam_receiver
    left_image_output: left_image_transmitter
    right_image_output: right_image_transmitter
    left_cam_output: left_cam_transmitter
    right_cam_output: right_cam_transmitter
    inputs: [left_image_receiver, right_image_receiver, left_cam_receiver, right_cam_receiver]
    outputs: [left_image_transmitter, right_image_transmitter, left_cam_transmitter, right_cam_transmitter]
- name: left_image_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: left_image_transmitter
    min_size: 1
- name: right_image_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: right_image_transmitter
    min_size: 1
- name: left_cam_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: left_cam_transmitter
    min_size: 1
- name: right_cam_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 12
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: right_cam_transmitter
    min_size: 1
---
name: left_roundrobin
components:
- name: data_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver
    min_size: 1
- type: nvidia::gxf::Broadcast
  parameters:
    source: data_receiver
    mode: RoundRobin
- name: data_transmitter_dla0
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- name: data_transmitter_dla1
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
---
name: right_roundrobin
components:
- name: data_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver
    min_size: 1
- type: nvidia::gxf::Broadcast
  parameters:
    source: data_receiver
    mode: RoundRobin
- name: data_transmitter_dla0
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- name: data_transmitter_dla1
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
---
name: disparity_roundrobin
components:
- name: data_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver
    min_size: 1
- type: nvidia::gxf::Broadcast
  parameters:
    source: data_receiver
    mode: RoundRobin
- name: data_transmitter_dla0
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter_dla0
    min_size: 1
- name: data_transmitter_dla1
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter_dla1
    min_size: 1
---
name: gather
components:
- name: input_dla0
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 6
    policy: 0
- name: input_dla1
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 6
    policy: 0
- type: nvidia::gxf::MultiMessageAvailableSchedulingTerm
  parameters:
    receivers: [input_dla0, input_dla1]
    min_size: 1
- name: output
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 12
- type: nvidia::gxf::Gather
  parameters:
    sink: output
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: output
    min_size: 1
---
name: bi3d_dla0
components:
- name: data_receiver_left
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 6
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver_left
    min_size: 1
- name: data_receiver_right
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 6
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver_right
    min_size: 1
- name: data_receiver_disparity
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 6
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver_disparity
    min_size: 1
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 6
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter
    min_size: 1
- name: pool
  type: nvidia::gxf::BlockMemoryPool
  parameters:
    storage_type: 1
    block_size: 2211840
    num_blocks: 12
- name: forward_pool
  type: nvidia::gxf::BlockMemoryPool
  parameters:
    storage_type: 0
    block_size: 256
    num_blocks: 12
- name: stream
  type: nvidia::gxf::CudaStreamPool
  parameters:
    dev_id: 0
    stream_flags: 1
    stream_priority: 0
    reserved_size: 1
    max_size: 1
- type: nvidia::isaac::Bi3DInference
  parameters:
    left_image_name: "left_image"
    right_image_name: "right_image"
    output_name: "disparity"
    pool: pool
    forward_pool: forward_pool
    stream_pool: stream
    left_image_receiver: data_receiver_left
    right_image_receiver: data_receiver_right
    output_transmitter: data_transmitter
    disparity_receiver: data_receiver_disparity

    image_type: "RGB_U8"
    pixel_mean: [0.0078431372549, 0.0078431372549, 0.0078431372549]
    normalization: [0.00392156862, 0.00392156862, 0.00392156862]
    standard_deviation: [-127.5, -127.5, -127.5]

    max_batch_size: 32
    input_layer_width: 960
    input_layer_height: 576
    model_input_type: "RGB_U8"

    featnet_engine_file_path: "featnet_engine_file_path"
    featnet_input_layers_name: ["input.1"]
    featnet_output_layers_name: ["97"]

    segnet_engine_file_path: "segnet_engine_file_path"
    segnet_input_layers_name: ["input.1"]
    segnet_output_layers_name: ["278"]

    engine_type: "DLA_CORE_0"
    apply_sigmoid: false
    apply_thresholding: true
    threshold: 0.0
    threshold_value_low: 0.0
    threshold_value_high: 1.0
    max_disparity_levels: 64
    disparity_values: [18]
- type: nvidia::gxf::MemoryAvailableSchedulingTerm
  parameters:
    allocator: pool
    min_blocks: 1
- type: nvidia::gxf::MemoryAvailableSchedulingTerm
  parameters:
    allocator: forward_pool
    min_blocks: 1
---
name: bi3d_dla1
components:
- name: data_receiver_left
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 6
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver_left
    min_size: 1
- name: data_receiver_right
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 6
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver_right
    min_size: 1
- name: data_receiver_disparity
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 6
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver_disparity
    min_size: 1
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 6
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter
    min_size: 1
- name: pool
  type: nvidia::gxf::BlockMemoryPool
  parameters:
    storage_type: 1
    block_size: 2211840
    num_blocks: 12
- name: forward_pool
  type: nvidia::gxf::BlockMemoryPool
  parameters:
    storage_type: 0
    block_size: 256
    num_blocks: 12
- name: stream
  type: nvidia::gxf::CudaStreamPool
  parameters:
    dev_id: 0
    stream_flags: 1
    stream_priority: 0
    reserved_size: 1
    max_size: 1
- type: nvidia::isaac::Bi3DInference
  parameters:
    left_image_name: "left_image"
    right_image_name: "right_image"
    output_name: "disparity"
    pool: pool
    forward_pool: forward_pool
    stream_pool: stream
    left_image_receiver: data_receiver_left
    right_image_receiver: data_receiver_right
    output_transmitter: data_transmitter
    disparity_receiver: data_receiver_disparity

    image_type: "RGB_U8"
    pixel_mean: [0.0078431372549, 0.0078431372549, 0.0078431372549]
    normalization: [0.00392156862, 0.00392156862, 0.00392156862]
    standard_deviation: [-127.5, -127.5, -127.5]

    max_batch_size: 32
    input_layer_width: 960
    input_layer_height: 576
    model_input_type: "RGB_U8"

    featnet_engine_file_path: "featnet_engine_file_path"
    featnet_input_layers_name: ["input.1"]
    featnet_output_layers_name: ["97"]

    segnet_engine_file_path: "segnet_engine_file_path"
    segnet_input_layers_name: ["input.1"]
    segnet_output_layers_name: ["278"]

    engine_type: "DLA_CORE_1"
    apply_sigmoid: false
    apply_thresholding: true
    threshold: 0.0
    threshold_value_low: 0.0
    threshold_value_high: 1.0
    max_disparity_levels: 64
    disparity_values: [18]
- type: nvidia::gxf::MemoryAvailableSchedulingTerm
  parameters:
    allocator: pool
    min_blocks: 1
- type: nvidia::gxf::MemoryAvailableSchedulingTerm
  parameters:
    allocator: forward_pool
    min_blocks: 1
---
name: sync_bi3d_dla0
components:
- name: rx
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 6
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: rx
    min_size: 1
- name: tx
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 6
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: tx
    min_size: 1
- type: nvidia::gxf::CudaStreamSync
  parameters:
    rx: rx
    tx: tx
---
name: sync_bi3d_dla1
components:
- name: rx
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 6
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: rx
    min_size: 1
- name: tx
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 6
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: tx
    min_size: 1
- type: nvidia::gxf::CudaStreamSync
  parameters:
    rx: rx
    tx: tx
---
name: bi3d_postprocess
components:
- name: pool
  type: nvidia::gxf::BlockMemoryPool
  parameters:
    storage_type: 1
    block_size: 2211840
    num_blocks: 24
- name: data_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver
    min_size: 1
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 12
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter
    min_size: 1
- type: nvidia::isaac::bi3d::Bi3DPostprocessor
  parameters:
    bi3d_receiver: data_receiver
    output_transmitter: data_transmitter
    pool: pool
    disparity_tensor_name: "disparity"
    disparity_values_tensor_name: "bi3d_inference_disparities"
    disparity_values: [18]
- type: nvidia::gxf::MemoryAvailableSchedulingTerm
  parameters:
    allocator: pool
    min_blocks: 1
---
name: disparity_compositor
components:
- name: disparity_in
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 2
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: disparity_in
    min_size: 1
- name: left_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 2
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: left_cam_receiver
    min_size: 1
- name: right_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 2
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: right_cam_receiver
    min_size: 1
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter
    min_size: 1
- type: nvidia::isaac::DisparityCompositor
  parameters:
    left_camera_model_receiver: left_cam_receiver
    right_camera_model_receiver: right_cam_receiver
    disparity_receiver: disparity_in
    output_transmitter: data_transmitter
    min_disparity: 0
    max_disparity: -1
---
name: bi3d_output_sink
components:
- name: input
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: input
    min_size: 1
- name: sink
  type: nvidia::isaac_ros::MessageRelay
  parameters:
    source: input
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: sync/left_image_transmitter
    target: left_roundrobin/data_receiver
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: sync/right_image_transmitter
    target: right_roundrobin/data_receiver
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: sync/left_cam_transmitter
    target: disparity_compositor/left_cam_receiver
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: sync/right_cam_transmitter
    target: disparity_compositor/right_cam_receiver
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: left_roundrobin/data_transmitter_dla0
    target: bi3d_dla0/data_receiver_left
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: right_roundrobin/data_transmitter_dla0
    target: bi3d_dla0/data_receiver_right
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: disparity_roundrobin/data_transmitter_dla0
    target: bi3d_dla0/data_receiver_disparity
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: disparity_roundrobin/data_transmitter_dla1
    target: bi3d_dla1/data_receiver_disparity
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: left_roundrobin/data_transmitter_dla1
    target: bi3d_dla1/data_receiver_left
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: right_roundrobin/data_transmitter_dla1
    target: bi3d_dla1/data_receiver_right
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: bi3d_dla0/data_transmitter
    target: sync_bi3d_dla0/rx
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: bi3d_dla1/data_transmitter
    target: sync_bi3d_dla1/rx
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: sync_bi3d_dla0/tx
    target: gather/input_dla0
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: sync_bi3d_dla1/tx
    target: gather/input_dla1
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: gather/output
    target: bi3d_postprocess/data_receiver
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: bi3d_postprocess/data_transmitter
    target: disparity_compositor/disparity_in
---
components:
- name: connection
  type: nvidia::gxf::Connection
  parameters:
    source: disparity_compositor/data_transmitter
    target: bi3d_output_sink/input
---
name: scheduler
components:
- name: clock
  type: nvidia::gxf::RealtimeClock
- type: nvidia::gxf::MultiThreadScheduler
  parameters:
    clock: clock
    worker_thread_number: 8
    stop_on_deadlock: false
